### DISCUSSION TITLE
TypeError: Promise.withResolvers is not a function

### DISCUSSION THREAD

**Discussion by jennabelle** [vercel/next.js#62122] [Help]
### Summary

Hello!

I am trying to migrate my app, from Next v12.3.4 to v13.5.6.  

I also installed `eslint-config-next@13.5.6` and removed the `next-transpile-modules` plugin, since that plugin is now available in Next 13. 

After running the `new-link` and `next-image-to-legacy-image` codemods, I noticed that I get this error when i try to run my app locally.

```
TypeError: Promise.withResolvers is not a function
    at Batcher.batch (/Users/jenna.vuong/mx-app/node_modules/next/dist/lib/batcher.js:39:54)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
```

Does anyone have any tips on how to resolve this error?  

It looks like the `Promise.withResolvers()` polyfill is being referenced, in the `next/dist/lib/batcher.js` file, but it's actually not getting that function, hence the error, `Promise.withResolvers()` is not a function

### Additional information

```js
Operating System:
  Platform: darwin
  Arch: arm64
  Version: Darwin Kernel Version 23.1.0: Mon Oct  9 21:32:11 PDT 2023; root:xnu-10002.41.9~7/RELEASE_ARM64_T6030
Binaries:
  Node: 20.10.0
  npm: 10.4.0
  Yarn: N/A
  pnpm: N/A
Relevant Packages:
  next: 13.5.6
  eslint-config-next: 13.5.6
  react: 18.2.0
  react-dom: 18.2.0
  typescript: 5.2.2
Next.js Config:
  output: N/A
```

**Comment by kodai3** [comment]
Maybe you should upgrade typescript to v5.4
https://github.com/microsoft/TypeScript/issues/56483

**Comment by axeleroy** [comment]
`Promise.withResolvers()` is [not available in versions of Node.js < 22](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/withResolvers#browser_compatibility), so you would have to update Node (or use a polyfill) in addition to upgrading to TypeScript 5.4

**Comment by momanamjad** [comment]
Got it! This is a Node.js version compatibility issue. Here's the fix:

**The Problem:**
`Promise.withResolvers()` is a relatively new JavaScript API that was only added in 
**Node.js 22.0.0**. Your machine is running Node 20.10.0, which doesn't have this method.

Next.js 13.5.6 uses this API internally in `batcher.js`, which is why you're getting 
this error.

**Solution - Upgrade Node.js:**

1. **Option A: Upgrade to Node.js 22 LTS (Recommended)**
```bash
   # Using nvm (Node Version Manager)
   nvm install 22
   nvm use 22
   
   # Verify
   node --version  # Should show v22.x.x
```

2. **Option B: Use Node.js 20 LTS (if you need to stay on 20)**
   - Upgrade Node.js 20 to the latest patch version:
```bash
   nvm install 20.15.0  # Latest Node 20
   nvm use 20.15.0
```
   - Then run `npm install` again and try building

3. **Option C: Use a Polyfill (Not recommended, but works)**
   Add this to your `next.config.js`:
```javascript
   if (!Promise.withResolvers) {
     Promise.withResolvers = function() {
       let resolve, reject;
       const promise = new Promise((res, rej) => {
         resolve = res;
         reject = rej;
       });
       return { promise, resolve, reject };
     };
   }
```

**Recommended Path:**
1. Delete `node_modules` and `package-lock.json`
```bash
   rm -rf node_modules package-lock.json
```

2. Upgrade Node.js to 22.x
```bash
   nvm install 22
   nvm use 22
```

3. Reinstall dependencies
```bash
   npm install
   npm run dev
```

This should resolve the issue! The Promise.withResolvers() API is much cleaner and safer, 
so upgrading to Node 22 is definitely the way to go.
